django/db/models/query.py

```diff
@@ -870,18 +868,17 @@ def annotate(self, *args, **kwargs):
         """
         annotations = OrderedDict()  # To preserve ordering of args
         for arg in args:
-            # The default_alias property may raise a TypeError, so we use
-            # a try/except construct rather than hasattr in order to remain
-            # consistent between PY2 and PY3 (hasattr would swallow
-            # the TypeError on PY2).
+            # The default_alias property may raise a TypeError. Use try/except
+            # instead of hasattr because hasattr may also raise TypeError if
+            # attr is not a string.
             try:
                 if arg.default_alias in kwargs:
                     raise ValueError("The named annotation '%s' conflicts with the "
                                      "default name for another annotation."
                                      % arg.default_alias)
-            except (AttributeError, TypeError):
+                annotations[arg.default_alias] = arg
```

Comment:
I don't see the need to move this into the try/except block. Generally we try to minimize what's in a try/except to make it clear what code is expected to raise the exception.  I think the old version in `aggregate()` with `arg.default_alias` only in the block is clearest.